File: backend/routers/breakeven.py
from datetime import datetime

import asyncio
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from lib.cache import get_cache, set_cache
from lib.fred import fred_series, fred_latest, n_days_ago

router = APIRouter()
CACHE_KEY = "breakeven"


@router.get("/breakeven")
async def breakeven():
    cached = get_cache(CACHE_KEY)
    if cached:
        return cached

    start = n_days_ago(365 * 2 + 30)

    try:
        us_raw, us_latest = await asyncio.gather(
            fred_series("T10YIE", start),
            fred_latest("T10YIE"),
            return_exceptions=True,
        )
        if isinstance(us_raw, Exception):
            us_raw = []
        if isinstance(us_latest, Exception):
            us_latest = None

        data = [{"date": p["date"], "us": p["value"]} for p in us_raw]
        current_us = us_latest["value"] if us_latest else None
        current_us_date = us_latest["date"] if us_latest else None

        result = {
            "data": data,
            "currentUS": current_us,
            "currentUSDate": current_us_date,
            "usAboveTarget": (current_us > 2.0) if current_us is not None else None,
            "norwayNote": (
                "Norwegian breakeven inflation: Norges Bank does not publish daily "
                "market-implied breakeven rates. Inflation-linked government bonds "
                "(NGB-IL) are thinly traded and not available via public API."
            ),
            "germanyNote": (
                "German breakeven inflation: ECB real yield data requires institutional "
                "access. No free daily series is available via FRED for German "
                "inflation-linked bonds."
            ),
            "updatedAt": datetime.utcnow().isoformat() + "Z",
        }
        set_cache(CACHE_KEY, result)
        return result
    except Exception as e:
        return JSONResponse(status_code=500, content={"error": str(e)})

File: src/pages/api/breakeven.ts
// Section 7: Breakeven Inflation
// US: T10YIE (10Y breakeven inflation, FRED)
// Germany: DFII10 is US real yield — for Germany, ECB data is not available without subscription.
//   We'll note this in the UI. We can use FRED BAMLHE00EHYIEY as a proxy check,
//   but cleanest is to use T10YIEM (if exists) — actually for Germany we'll note limitation.
// Norway: Norges Bank publishes inflation expectations surveys, not daily market breakevens.
//   No daily Norwegian breakeven series available — we note this clearly.

import type { NextApiRequest, NextApiResponse } from "next";
import { fredSeries, fredLatest, nDaysAgo } from "@/lib/fred";
import { getCache, setCache } from "@/lib/cache";

const CACHE_KEY = "breakeven";

export interface BreakevenPoint {
  date: string;
  us: number | null;
}

export interface BreakevenResponse {
  data: BreakevenPoint[];
  currentUS: number | null;
  currentUSDate: string | null;
  usAboveTarget: boolean | null;
  norwayNote: string;
  germanyNote: string;
  updatedAt: string;
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<BreakevenResponse | { error: string }>
) {
  const cached = getCache<BreakevenResponse>(CACHE_KEY);
  if (cached) return res.status(200).json(cached);

  const start = nDaysAgo(365 * 2 + 30);

  try {
    const [usRaw, usLatest] = await Promise.all([
      fredSeries("T10YIE", start).catch(() => []),
      fredLatest("T10YIE").catch(() => null),
    ]);

    const data: BreakevenPoint[] = usRaw.map((p) => ({
      date: p.date,
      us: parseFloat(p.value),
    }));

    const result: BreakevenResponse = {
      data,
      currentUS: usLatest?.value ?? null,
      currentUSDate: usLatest?.date ?? null,
      usAboveTarget: usLatest ? usLatest.value > 2.0 : null,
      norwayNote:
        "Norwegian breakeven inflation: Norges Bank does not publish daily market-implied breakeven rates. Inflation-linked government bonds (NGB-IL) are thinly traded and not available via public API.",
      germanyNote:
        "German breakeven inflation: ECB real yield data requires institutional access. No free daily series is available via FRED for German inflation-linked bonds.",
      updatedAt: new Date().toISOString(),
    };

    setCache(CACHE_KEY, result);
    res.status(200).json(result);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
}
